cpp/src/strings/regex/regcomp.cpp

```diff
@@ -0,0 +1,1180 @@
+                }
+                if (*exprp != '}')
+                {
+                    exprp = exprp_backup;
+                    break;
+                }
+                if (buff[0] != 0)
+                    sscanf(buff, "%hd", &yy_max_count);
+            }
+            exprp++;
+            if (*exprp == '?')
+            {
+                exprp++;
+                return COUNTED_LAZY;
+            }
+            return COUNTED;
+        }
+        case '|':
+            return OR;
+        case '.':
+            return dot_type;
+        case '(':
+            if (*exprp == '?' && *(exprp + 1) == ':')  // non-capturing group
+            {
+                exprp += 2;
+                return LBRA_NC;
+            }
+            return LBRA;
+        case ')':
+            return RBRA;
+        case '^':
+            return BOL;
+        case '$':
+            return EOL;
+        case '[':
+            return bldcclass();
+        }
+        return CHAR;
+    }
+public:
+    struct Item
+    {
+        int t;
+        union
+        {
+            char32_t yy;
+            int yyclass_id;
+            struct
+            {
+                short n;
+                short m;
+            } yycount;
+        } d;
+    };
+    std::vector<Item> m_items;
+
+    bool m_has_counted;
+
+    RegParser(const char32_t* pattern, int dot_type, reprog& prog) : m_prog(prog)
+    {
+        exprp = pattern;
+        lexdone = false;
+        m_has_counted = false;
+        int token;
+        while ((token = lex(dot_type)) != END)
+        {
+            Item item;
+            item.t = token;
+            if (token == CCLASS || token == NCCLASS)
+                item.d.yyclass_id = yyclass_id;
+            else if (token == COUNTED || token == COUNTED_LAZY)
+            {
+                item.d.yycount.n = yy_min_count;
+                item.d.yycount.m = yy_max_count;
+                m_has_counted = true;
+            }
+            else
+                item.d.yy = yy;
+            m_items.push_back(item);
+        }
+    }
+};
+
+/**
+ * @brief The compiler converts class list into instructions.
+ */
+class RegCompiler
+{
+    reprog& m_prog;
+
+    struct Node
+    {
+        int id_first;
+        int id_last;
+    };
+
+    int cursubid;
+    int pushsubid;
+    std::vector<Node> andstack;
+
+    struct Ator
+    {
+        int t;
+        int subid;
+    };
+
+    std::vector<Ator> atorstack;
+
+    bool lastwasand;
+    int	nbra;
+
+    inline void pushand(int f, int l)
+    {
+        andstack.push_back({ f, l });
+    }
+
+    inline Node popand(int op)
+    {
+        if( andstack.size() < 1 )
+        {
+            //missing operand for op
+            int inst_id = m_prog.add_inst(NOP);
+            pushand(inst_id, inst_id);
+        }
+        Node node = andstack[andstack.size() - 1];
+        andstack.pop_back();
+        return node;
+    }
+
+    inline void pushator(int t)
+    {
+        Ator ator;
```

Comment:
> Ator? Allocator? no need to abbreviate here

I figured out that the code keeps a list of `operators` and `operands` and so manages a vector of `ands` and `ators` since both are `opers`. Since `operator` is a reserved word, I need to come up with a different name here for both. Would it be ok to push this off to #3582 ?

Reply:
Sounds good. I will ponder on some better names, definitely we can do better than `ators` and `ands`.